[code-infra] Add options to docs-api generation to support X code structure

alexfauquette opened the pull request:
PR created in parallel of https://github.com/mui/mui-x/pull/11303

### Support the `@see` JSDocs descriptor

The data grid and pickers team use it as follow:
```
* @see See the {@link https://mui.com/x/react-date-pickers/timezone/ timezones documentation} for more details.
```
  The script extract the link url and text in the json with data and the sentence is saved in the translation file with `{{link}}` to indcate where the link should be placed.

### Customize import docs data

The config object now have two distinct parametters:
- `translationPagesDirectory`: the path used to generate translation files
- `importTranslationPagesDirectory`: the path used in generated files

The reason is that in X repo, `docs/` is an alias for `@monorepo/docs` and we use `docsx` to link to the `docs` folder (I know it's weird) so we need them to be distinct:
- `translationPagesDirectory: "docs/translations/charts"`
- `importTranslationPagesDirectory: "docsx/translations/charts"`

### Add `propsWithoutDefaultVerification`

An array of props that will not be verified.
The script make sure that if you use `const { debounceMs = rootProps.filterDebounceMs } = props` The the JSDocs should be `@default rootProps.filterDebounceMs` That's useless for users. So I keep the default set in the `@default` and allow to disable the warning.

It's used by
- data grid
	- debounceMs
	- quickFilterParser
- pickers
	- 'dateIcon',
	- 'hidden',
	- 'timeIcon',
	- 'toolbarPlaceholder',
	- 'ampm',
	- 'format',
	- 'changeImportance',
	- 'desktopModeMediaQuery',

### Fix `removeOutdatedApiDocsTranslations`

The script in charge of removing all the files that were not supposed to be in the folder was not up to date.

It was cleaning all the folders into `docs/translations/api-docs` (where material docs is)

Now it takes a list of folders to clean, which are generated from the `translationPagesDirectory` of projects.

### `getComponentImports`

You can now override how components import are generated

### `generateJsonFileOnly`

That's an option only used by base. Added to the settings to remove as much as possible lines that check "if project name is based, do that, otherwise do this" 

### `skipAnnotatingComponentDefinition`

The script is annotating components with links to demo, and api pages. Since the X team don't have a one to one relation ship between components, demo, and api pages. I disable this annotation to move that for latter.

Also used by base

### Generate default when not defaultized

If the props were not explicitly defined, no default were assigned.

I removed this `if (props !== undefined)` condition, because X has a lot of props that don't respect this `const { ... } = props` pattern.

A consequence for core is the addition of some `default: false` that were not reported before.

### Generated js files

Add a `/\.\/` in the `getInitialProps` because without it, Next was complaining on X

mnajdova commented on packages/api-docs-builder/utils/createDescribeableProp.ts:
```diff
@@ -8,6 +8,17 @@ export interface DescribeablePropDescriptor {
 
+export type CreateDescribeablePropSettings = {
+  /**
+   * name of props which do not require default annotation.
```
```suggestion
   * Names of props that do not require default annotation.
```

mnajdova commented on packages/api-docs-builder/utils/createDescribeableProp.ts:
```diff
@@ -8,6 +8,17 @@ export interface DescribeablePropDescriptor {
+   */
+  propsWithoutMandatoryDefaultAnnotation?: string[];
+  /**
+   * name of props which do not check the annotation equal runtime default.
```
```suggestion
   * Names of props that do not check the annotation equal runtime default.
```

mnajdova commented on packages/api-docs-builder/utils/generateApiTranslation.ts:
```diff
@@ -0,0 +1,52 @@
+  reactApi: ReactApi,
+  languages: string[],
+) {
+  const hookName = reactApi.name;
```
```suggestion
  const definitionName = reactApi.name;
```

It can be both hook and component.

mnajdova commented on packages/api-docs-builder/utils/generateApiTranslation.ts:
```diff
@@ -0,0 +1,52 @@
+  translations: object;
+}
+
+export default function generateApiTranslations<ReactApi extends MinimalReactAPI>(
```
👍 for reusing the same function, not sure why I added a duplicate.

mnajdova commented on docs/translations/api-docs-base/dropdown-menu/dropdown-menu.json:
```diff
@@ -1,9 +0,0 @@
-{
```
Why do we have removed files?

alexfauquette replied:
The cleaning method was to delete all sub-folders of `docs/translation/api-docs` that are not generated.

Since base and joy are in distinct folder they were not concerned by the cleaning strategy. Those files got removed when I adapted the strategy to work with any folder provided in the setting.

I did not found those file imported anywhere. I assume it's some components name that got modified without removing the auto-generated API

alexfauquette commented:
@mnajdova @michaldudak I added last modification to support some other features of the data grid:

The API translation path was not recognized on X: I fixed it with the following commit: https://github.com/mui/material-ui/pull/40111/commits/45cd0901c43a1fc9f0a4bc6a5b788473e821271d

The grid and pickers use some `@see` JSDocs decorator to link to dedicated docs pages. I added the script that extract the data and render them in the API pages

michaldudak commented on docs/public/_redirects:
```diff
@@ -510,19 +510,17 @@ https://v4.material-ui.com/* https://v4.mui.com/:splat 301!
 /toolpad/* https://mui-toolpad-docs.netlify.app/toolpad/:splat 200
 /r/toolpad-* https://mui-toolpad-docs.netlify.app/r/toolpad-:splat 200
 
-### MUI Design kits
```
Is this related to API docs generation?

alexfauquette replied:
Probably come from the moment I mixed between next/master branches. I'm reverting this

alexfauquette commented on docs/public/_redirects:
```diff
@@ -510,19 +510,17 @@ https://v4.material-ui.com/* https://v4.mui.com/:splat 301!
-/:lang/x/* https://material-ui-x.netlify.app/:lang/x/:splat 200
+/static/x/* https://docs-next--material-ui-x.netlify.app/static/x/:splat 200
+/static/* https://mui.com/static/:splat 200
+/x/_next/* https://docs-next--material-ui-x.netlify.app/_next/:splat 200
+/x/* https://docs-next--material-ui-x.netlify.app/x/:splat 200
+/r/x-* https://docs-next--material-ui-x.netlify.app/r/x-:splat 200
+/:lang/x/* https://docs-next--material-ui-x.netlify.app/:lang/x/:splat 200
 
 ## MUI Toolpad
 /static/toolpad/* https://mui-toolpad-docs.netlify.app/static/toolpad/:splat 200
 /toolpad/_next/* https://mui-toolpad-docs.netlify.app/_next/:splat 200
 /toolpad/* https://mui-toolpad-docs.netlify.app/toolpad/:splat 200
 /r/toolpad-* https://mui-toolpad-docs.netlify.app/r/toolpad-:splat 200
 
-### MUI Design kits
-/figma/getting-started/ https://mui-org.notion.site/MUI-for-Figma-docs-4eb22360388047a9a31fc19d300a285e 302
-/r/material-ui-figma-* https://mui-store.netlify.app/r/material-ui-figma-:splat 200
-/r/joy-ui-figma-* https://mui-store.netlify.app/r/joy-ui-figma-:splat 200
+/* https://mui.com/:splat 301!
```
```suggestion
## Unlike the store that expect to be hosted under a subfolder,
## MUI X is configured to be hosted at the root.
/static/x/* https://material-ui-x.netlify.app/static/x/:splat 200
/x/_next/* https://material-ui-x.netlify.app/_next/:splat 200
/x/* https://material-ui-x.netlify.app/x/:splat 200
/r/x-* https://material-ui-x.netlify.app/r/x-:splat 200
/:lang/x/* https://material-ui-x.netlify.app/:lang/x/:splat 200

## MUI Toolpad
/static/toolpad/* https://mui-toolpad-docs.netlify.app/static/toolpad/:splat 200
/toolpad/_next/* https://mui-toolpad-docs.netlify.app/_next/:splat 200
/toolpad/* https://mui-toolpad-docs.netlify.app/toolpad/:splat 200
/r/toolpad-* https://mui-toolpad-docs.netlify.app/r/toolpad-:splat 200

### MUI Design kits
/figma/getting-started/ https://mui-org.notion.site/MUI-for-Figma-docs-4eb22360388047a9a31fc19d300a285e 302
/r/material-ui-figma-* https://mui-store.netlify.app/r/material-ui-figma-:splat 200
/r/joy-ui-figma-* https://mui-store.netlify.app/r/joy-ui-figma-:splat 200
```

michaldudak commented:
Could you please add a brief description to the PR of what this change is about and how it can be tested?

michaldudak reviewed (approved):
The link text should be translatable (similarly to CSS class descriptions).
But since this is not critical right now as we don't have translations, you can do it in a separate PR, so this one doesn't get stale and can unblock other work.
Everything else looks fine to me!
